### PR TITLE
Revert change (#90674)

### LINKED ISSUE #103011: Significant ingest performance regression while handling parsing dynamically added objects
### Elasticsearch Version

8.11+

### Installed Plugins

_No response_

### Java Version

_bundled_

### OS Version

any

### Problem Description

While indexing documents with dynamically added fields, if multiple of those fields are objects are are deeper than 3-5 objects, performance is degraded significantly.

This bug was introduced in: https://github.com/elastic/elasticsearch/pull/90674

### Steps to Reproduce

I built a micro-benchmark to replicate this particular bug. 

But, any dynamic object parsing will cause this.

To mitigate the issue, statically mapped fields for objects is required. 

### Logs (if relevant)

Attached are two JFR files indicating the difference between when the bug was introduce vs. not.
[Archive 2.zip](https://github.com/elastic/elasticsearch/files/13569248/Archive.2.zip)

### PR DESCRIPTION
**Pull request by benwtrent**
Reverts https://github.com/elastic/elasticsearch/pull/90674

The revert is not perfectly clean as there are some minor adjustments to account for later changes.

This is in contrast with: https://github.com/elastic/elasticsearch/pull/103858

closes: https://github.com/elastic/elasticsearch/issues/103011

### CODE REVIEW

**Review by javanna — APPROVED** [review]
I went through this and compared line by line with the PR that it reverts. It looks good to me. I think it is wise to revert at this point, and reconsider what we want to do next. You mentioned it did not revert cleanly: did the problems only come from the float handling in `postProcessDynamicArrayMapping` for dynamic mapping of vector fields?

Thanks for jumping on this!

**Comment by benwtrent** [comment]
> did the problems only come from the float handling in postProcessDynamicArrayMapping for dynamic mapping of vector fields?

Correct, the logic there changed from a separate bug fix.

**Comment by benwtrent** [comment]
@elasticmachine update branch

**Comment by benwtrent** [comment]
@elasticmachine update branch

File: server/src/main/java/org/elasticsearch/index/mapper/DocumentParserContext.java [hunk 1]
```diff
@@ -166,9 +166,9 @@ protected DocumentParserContext(
             mappingParserContext,
             source,
             new HashSet<>(),
-            new LinkedHashMap<>(),
+            new HashMap<>(),
             new HashSet<>(),
-            new LinkedHashMap<>(),
+            new HashMap<>(),
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/index/mapper/DocumentParserContext.java:171]: ++ it makes sense to go back to the original hashmap, which we had before.

**Comment by benwtrent** [comment]
run elasticsearch-ci/part-2

**Comment by benwtrent** [comment]
run elasticsearch-ci/part-1
run elasticsearch-ci/part-3
run elasticsearch-ci/part-4
run elasticsearch-ci/bwc-snapshots
run elasticsearch-ci/packaging-tests-unix-sample
run elasticsearch-ci/packaging-tests-windows-sample

**Comment by felixbarny** [comment]
FWIW, I'm also happy about this change as it will make #102936 and #96235 much easier as we don't need separate implementations for `mapperSize` in both `Mapper` and `Mapper.Builder` that need to be kept in sync.

Before we re-introduce this change (in case we want to do that at some point), we should really merge the two PRs mentioned above first.

**Comment by benwtrent** [comment]
@felixbarny 
> Before we re-introduce this change (in case we want to do that at some point), we should really merge the two PRs mentioned above first.

By "re-introduce" you mean adding back "keep builders" change that this PR reverts?

**Comment by benwtrent** [comment]
@elasticmachine update branch

**Comment by felixbarny** [comment]
> By "re-introduce" you mean adding back "keep builders" change that this PR reverts?

Yes, that's what I meant. Aka revert the revert.

So before we're considering to re-introduce to "Store dynamic mapping updates as builders", we should first merge https://github.com/elastic/elasticsearch/pull/102936 and https://github.com/elastic/elasticsearch/pull/96235. We can then weigh in if the additional complexity of maintaining a `mapperSize` implementation in both `Mapper` and `Mapper.Builder` is worth the optimization.

**Pull request merged**